File: C++/BehaviorTreesLib/RepeatUntilFail.cpp
#include "RepeatUntilFail.h"

namespace fluentBehaviorTree
{
	Node * RepeatUntilFail::copy()
	{
		RepeatUntilFail* newNode = new RepeatUntilFail(this->getName());
		newNode->addChild(((this->mChild)->copy()));
		return newNode;
	}
	RepeatUntilFail::RepeatUntilFail(std::string name)
	{
		this->setName(name);
	}

	RepeatUntilFail::~RepeatUntilFail()
	{
		if (mChild != nullptr)
		{
			delete mChild;
			mChild = nullptr;
		}
	}

	// Repeat until FAILURE
	EStatus RepeatUntilFail::tickNode()
	{
		this->setResult(mChild->tick());

		if (this->getResult() == EStatus::SUCCESS) this->setResult(EStatus::RUNNING);
		else if (this->getResult() == EStatus::FAILURE) this->setResult(EStatus::SUCCESS);

		return this->getResult();
	}
}

File: C#/BehaviorTree/Branch/Decorator/RepeatUntilFail.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace FluentBehaviorTree
{
    public class RepeatUntilFail : Decorator
    {
        public RepeatUntilFail(string name)
        {
            this.Name = name;
        }

        public override object Clone()
        {
            RepeatUntilFail newNode = new RepeatUntilFail(this.Name);
            newNode.AddChild(this.child.Clone() as Node);
            return newNode;
        }

        /// <summary>
        /// Repeat until FAILURE
        /// </summary>
        /// <returns></returns>
        protected override Status tick()
        {
            this.Result = child.Tick();

            if (this.Result == Status.SUCCESS) this.Result = Status.RUNNING;
            else if (this.Result == Status.FAILURE) this.Result = Status.SUCCESS;

            return this.Result;
        }
    }
}
